SOURCES/etc/xapi.d/plugins/sdncontroller.py

```diff
@@ -0,0 +1,320 @@
+
+    def parse_direction(self):
+        direction = self.args.get("direction")
+
+        if direction is None:
+            log_and_raise_error(E_PARSER, "direction parameter is missing")
+
+        dir = direction.lower().split("/")
+        has_to = "to" in dir
+        has_from = "from" in dir
+
+        if not (has_to or has_from):
+            log_and_raise_error(E_PARSER, "'{}' is not a valid direction".format(direction))
+
+        return (has_to, has_from)
+
+    def parse_protocol(self):
+        protocol = self.args.get("protocol")
+
+        if protocol is None:
+            log_and_raise_error(E_PARSER, "protocol parameter is missing")
+
+        protocol = protocol.lower()
+
+        if protocol not in VALID_PROTOCOLS:
+            log_and_raise_error(E_PARSER, "'{}' is not a supported protocol".format(protocol))
+
+        return protocol
+
+    def parse_port(self):
+        port = self.args.get("port")
+        if port is None:
+            return None
+
+        try:
+            p = int(port)
+            if not (0 <= p <= 65535):
+                raise ValueError
+            return port
+        except ValueError:
+            log_and_raise_error(E_PARSER, "'{}' is not a valid port".format(port))
+
+    def parse_allow(self):
+        allow = self.args.get("allow")
+        if allow is None:
+            log_and_raise_error(E_PARSER, "allow parameter is missing")
+
+        if allow.lower() not in ["true", "false"]:
+            log_and_raise_error(E_PARSER, "allow parameter should be true or false, not '{}'".format(allow))
+
+        if allow.lower() == "true":
+            return True
+
+        return False
+
+    def parse_priority(self):
+        priority = self.args.get("priority")
+        if priority is None:
+            return None
+
+        try:
+            p = int(priority)
+            if not (0 <= p <= 65535):
+                raise ValueError
+            return priority
+        except ValueError:
+            log_and_raise_error(E_PARSER, "'{}' is not a valid priority".format(priority))
+
+    def read(self, key, parse_fn, dests=None):
+        # parse_fn can return a single value or a tuple of values.
+        # In this case we are expecting dests to match the expected
+        # returned tuple
+        val = parse_fn()
+
+        if dests is not None:
+            if not isinstance(val, tuple):
+                log_and_raise_error(
+                    E_PARSER,
+                    "Internal error: parse {}: doesn't return a tuple while dests is set".format(key)
+                )
+
+            if len(dests) != len(val):
+                log_and_raise_error(
+                    E_PARSER,
+                    "Internal error: parse {}: dests is {} while {} was expected".format(key, len(dests), len(val))
+                )
+
+            for d, v in zip(dests, val):
+                self.values[d] = v
+
+            return self
+
+        self.values[key] = val
+        return self
```

Comment:
Last little thing, I don't see any usage of the return value of `Parser.read()` so I don't think it need to return itself.